hibernate-core/src/test/java/org/hibernate/orm/test/version/UserVersionTest.java

```diff
@@ -0,0 +1,201 @@
+					session.persist( testEntity );
+				}
+		);
+
+		inTransaction(
+				session -> {
+					TestEntity testEntity = session.get( TestEntity.class, 1 );
+					assertThat( testEntity.getVersion().getRev(), CoreMatchers.is( 1l ) );
+				}
+		);
+	}
+
+
+	@Entity(name = "TestEntity")
+	public static class TestEntity {
+		@Id
+		public int id;
+
+		@Version
+		@Column(name = "revision")
+		CustomVersion version;
+
+		public TestEntity() {
+		}
+
+		public TestEntity(int id) {
+			this.id = id;
+		}
+
+		public int getId() {
+			return id;
+		}
+
+		public CustomVersion getVersion() {
+			return version;
+		}
+	}
+
+	public static class CustomVersion implements Serializable {
+		private final long rev;
+
+
+		public CustomVersion(long rev) {
+			this.rev = rev;
+		}
+
+		public long getRev() {
+			return rev;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( o == null || getClass() != o.getClass() ) {
+				return false;
+			}
+			CustomVersion that = (CustomVersion) o;
+			return rev == that.rev;
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash( rev );
+		}
+	}
+
+	public static class CustomVersionUserVersionType implements UserVersionType<CustomVersion> {
+
+		@Override
+		public int getSqlType() {
+			return Types.BIGINT;
+		}
+
+		@Override
+		public Class<CustomVersion> returnedClass() {
+			return CustomVersion.class;
+		}
+
+		@Override
+		public boolean equals(CustomVersion x, CustomVersion y) throws HibernateException {
+			return x.equals( y );
+		}
+
+		@Override
+		public int hashCode(final CustomVersion x) throws HibernateException {
+			return x.hashCode();
+		}
+
+		@Override
+		public CustomVersion nullSafeGet(
+				ResultSet resultSet,
+				int position,
+				SharedSessionContractImplementor session,
+				Object owner) throws HibernateException, SQLException {
+			return new CustomVersion( resultSet.getLong( position ) );
+		}
+
+		@Override
+		public void nullSafeSet(
+				PreparedStatement statement,
+				CustomVersion value,
+				int index,
+				SharedSessionContractImplementor session) throws HibernateException, SQLException {
+			statement.setLong( index, value.getRev() );
+		}
+
+		@Override
+		public CustomVersion deepCopy(CustomVersion value) throws HibernateException {
+			return value;
+		}
+
+		@Override
+		public boolean isMutable() {
+			return false;
+		}
+
+		@Override
+		public Serializable disassemble(CustomVersion value) {
+			return value;
+		}
+
+		@Override
+		public CustomVersion assemble(Serializable cached, Object owner) throws HibernateException {
+			return (CustomVersion) cached;
+		}
+
+		@Override
+		public CustomVersion replace(CustomVersion original, CustomVersion target, Object owner)
+				throws HibernateException {
+			return original;
+		}
+
+		@Override
+		public CustomVersion seed(SharedSessionContractImplementor session) {
+			return new CustomVersion( ++versionValue );
```

Comment:
seems the `versionValue` is only used in this entity class, so relocating the variable declaration within this class makes more sense. not a big deal though.